ethereum/blockwatch/block_watcher.go

```diff
@@ -185,46 +194,82 @@ func (w *Watcher) GetAllRetainedBlocks() ([]*miniheader.MiniHeader, error) {
+// syncChain syncs our local state of the chain to the latest block found via
+// Ethereum RPC
+func (w *Watcher) syncChain() error {
+	latestHeader, err := w.client.HeaderByNumber(nil)
 	if err != nil {
 		return err
 	}
-	if latestHeader == nil {
-		if w.startBlockDepth == rpc.LatestBlockNumber {
-			nextBlockNumber = nil // Fetch latest block
-		} else {
-			nextBlockNumber = big.NewInt(int64(w.startBlockDepth))
-		}
-	} else {
-		nextBlockNumber = big.NewInt(0).Add(latestHeader.Number, big.NewInt(1))
-	}
-	nextHeader, err := w.client.HeaderByNumber(nextBlockNumber)
+	lastStoredHeader, err := w.stack.Peek()
 	if err != nil {
-		if err == ethereum.NotFound {
-			log.WithFields(log.Fields{
-				"blockNumber": nextBlockNumber,
-			}).Trace("block header not found")
-			return nil // Noop and wait next polling interval
-		}
 		return err
 	}
 
-	events := []*Event{}
-	events, err = w.buildCanonicalChain(nextHeader, events)
-	// Even if an error occurred, we still want to emit the events gathered since we might have
-	// popped blocks off the Stack and they won't be re-added
-	if len(events) != 0 {
-		w.blockFeed.Send(events)
+	var numBlocksToFetch int
+	// No blocks stored yet, fetch the first
+	if lastStoredHeader == nil {
+		numBlocksToFetch = 1
+	} else {
+		// Noop if already caught up or ahead of latest block returned from Ethereum node
+		if latestHeader.Number.Int64() <= lastStoredHeader.Number.Int64() {
+			return nil
+		}
+		numBlocksToFetch = int(latestHeader.Number.Int64() - lastStoredHeader.Number.Int64())
 	}
-	if err != nil {
-		return err
+
+	allEvents := []*Event{}
+	var syncErr error
+	for i := 0; i < numBlocksToFetch; i++ {
+		// Optimization: If numBlocksToFetch is 1, we already know what the nextHeader is, so avoid
+		// fetching it again. If there is more then 1 block to fetch, compute each from the last
+		// stored and fetch it
+		nextHeader := latestHeader
+		if numBlocksToFetch != 1 {
+			lastStoredHeader, syncErr := w.stack.Peek()
+			if syncErr != nil {
+				break
+			}
+			nextBlockNumber := big.NewInt(0).Add(lastStoredHeader.Number, big.NewInt(1))
+			nextHeader, syncErr = w.client.HeaderByNumber(nextBlockNumber)
+			if syncErr != nil {
+				break
+			}
+		}
+
+		var events []*Event
+		events, syncErr = w.buildCanonicalChain(nextHeader, events)
+		allEvents = append(allEvents, events...)
+		if syncErr != nil {
+			break
+		}
+	}
+	if len(allEvents) == 0 {
+		return syncErr
+	}
+	if w.shouldRevertChanges(lastStoredHeader, allEvents) {
+		if err := w.stack.Reset(); err != nil {
+			return err
+		}
+	} else {
+		if err := w.stack.Checkpoint(); err != nil {
+			return err
+		}
+		w.blockFeed.Send(allEvents)
 	}
-	return nil
+
+	return syncErr
+}
+
+func (w *Watcher) shouldRevertChanges(lastStoredHeader *miniheader.MiniHeader, events []*Event) bool {
+	if len(events) == 0 || lastStoredHeader == nil {
+		return false
+	}
+	// If we haven't progressed in terms of block number, revert back to previous "latest" block.
```

Comment:
Can you elaborate on this? Shouldn't re-orgs only ever occur if one chain reached a higher block number than the one we were previously considering the "latest"?

Reply:
That is correct. However, syncing to the latest block will involve multiple ETH RPC calls and if any of them fail, we reach this part of the code, and so we check that if we haven't progressed higher before the failure happened, we revert the changes made.

E.g. 1, found 4 blocks are missing, fetched first block with logs, then second, on fetching the 3rd block we received a non-200 response. Since two blocks were added, we don't revert.

E.g. 2, found block re-org, removed a block, tried to fetch replacement block but got non-200 response, since we only removed a block at this point, we are not further ahead so revert the changes made.